File: 394-decode-string/decode-string.cpp
class Solution {
public:
    string decodeString(string s) {
        stack<char> st;
        for (int i = 0; i < s.size(); i++) {
            if (s[i] != ']') {
                st.push(s[i]);
            } else {
                string curr = "";
                while (st.top() != '[') {
                    curr = st.top() + curr;
                    st.pop();
                }

                st.pop();
                string number = "";

                while (!st.empty() && isdigit(st.top())) {
                    number = st.top() + number;
                    st.pop();
                }
                int n = stoi(number);
                while (n--) {
                    for (int p = 0; p < curr.size(); p++) {
                        st.push(curr[p]);
                    }
                }
            }
        }
        s = "";
        while(!st.empty()){
            s = st.top() + s;
            st.pop();
        }
        return s;
    }
};

File: 394-decode-string/decode-string.java
class Solution {
    public String decodeString(String s) {
        Stack<Integer> countStack = new Stack<>();
        Stack<StringBuffer> stringStack = new Stack<>();
        StringBuffer currString = new StringBuffer();
        int k = 0;
        for (char c : s.toCharArray()) {
            if (Character.isDigit(c)) {
                k = k * 10 + (c - '0');
            } else if (c == '[') {
                countStack.add(k);
                stringStack.add(currString);
                k = 0;
                currString = new StringBuffer();
            } else if (c == ']') {
                String inner = currString.toString();
                int count = countStack.pop();
                StringBuffer prev = stringStack.pop();
                for (int i = 0; i < count; i++) {
                    prev.append(inner);
                }
                currString = prev;
            } else {
                currString.append(c);
            }
        }

        return currString.toString();

    }
}